core/src/repair/quic_endpoint.rs

```diff
@@ -0,0 +1,624 @@
+    pub(crate) remote_address: SocketAddr,
+    pub(crate) bytes: Vec<u8>,
+    pub(crate) response_sender: Option<OneShotSender<Vec<Vec<u8>>>>,
+}
+
+#[derive(Error, Debug)]
+#[allow(clippy::enum_variant_names)]
+pub(crate) enum Error {
+    #[error(transparent)]
+    BincodeError(#[from] bincode::Error),
+    #[error(transparent)]
+    CertificateError(#[from] RcgenError),
+    #[error(transparent)]
+    ConnectError(#[from] ConnectError),
+    #[error(transparent)]
+    ConnectionError(#[from] ConnectionError),
+    #[error("Channel Send Error")]
+    ChannelSendError,
+    #[error("Invalid Identity: {0:?}")]
+    InvalidIdentity(SocketAddr),
+    #[error(transparent)]
+    IoError(#[from] IoError),
+    #[error("No Response Received")]
+    NoResponseReceived,
+    #[error(transparent)]
+    ReadToEndError(#[from] ReadToEndError),
+    #[error(transparent)]
+    WriteError(#[from] WriteError),
+    #[error(transparent)]
+    TlsError(#[from] rustls::Error),
+}
+
+#[allow(clippy::type_complexity)]
+pub(crate) fn new_quic_endpoint(
+    runtime: &tokio::runtime::Handle,
+    keypair: &Keypair,
+    socket: UdpSocket,
+    address: IpAddr,
+    remote_request_sender: Sender<RemoteRequest>,
+) -> Result<(Endpoint, AsyncSender<LocalRequest>, AsyncTryJoinHandle), Error> {
+    let (cert, key) = new_self_signed_tls_certificate(keypair, address)?;
+    let server_config = new_server_config(cert.clone(), key.clone())?;
+    let client_config = new_client_config(cert, key)?;
+    let mut endpoint = {
+        // Endpoint::new requires entering the runtime context,
+        // otherwise the code below will panic.
+        let _guard = runtime.enter();
+        Endpoint::new(
+            EndpointConfig::default(),
+            Some(server_config),
+            socket,
+            Arc::new(TokioRuntime),
+        )?
+    };
+    endpoint.set_default_client_config(client_config);
+    let cache = Arc::<RwLock<ConnectionCache>>::default();
+    let (client_sender, client_receiver) = tokio::sync::mpsc::channel(CLIENT_CHANNEL_CAPACITY);
+    let server_task = runtime.spawn(run_server(
+        endpoint.clone(),
+        remote_request_sender.clone(),
+        cache.clone(),
+    ));
+    let client_task = runtime.spawn(run_client(
+        endpoint.clone(),
+        client_receiver,
+        remote_request_sender,
+        cache,
+    ));
+    let task = futures::future::try_join(server_task, client_task);
+    Ok((endpoint, client_sender, task))
+}
+
+pub(crate) fn close_quic_endpoint(endpoint: &Endpoint) {
+    endpoint.close(
+        CONNECTION_CLOSE_ERROR_CODE_SHUTDOWN,
+        CONNECTION_CLOSE_REASON_SHUTDOWN,
+    );
+}
+
+fn new_server_config(cert: Certificate, key: PrivateKey) -> Result<ServerConfig, rustls::Error> {
+    let mut config = rustls::ServerConfig::builder()
+        .with_safe_defaults()
+        .with_client_cert_verifier(Arc::new(SkipClientVerification {}))
+        .with_single_cert(vec![cert], key)?;
```

Comment:
Need to turn the "retry" token to true on the server config to avoid the costless initial packets flood attack.